Add `LocalFilesystemToGoogleDriveOperator`

ulsc opened the pull request:
Operator to upload a list of files to a Google Drive folder, alongside a file which can be used to have more Google Drive operators.

ulsc committed:
Add `GoogleDriveUploadOperator`

ulsc committed:
Add tests for `GoogleDriveUploadOperator`

eladkal commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,116 @@
+        delete: bool = False,
+        chunk_size: int = 100 * 1024 * 1024,
+        resumable: bool = False,
+        api_version: str = 'v3',
```
Must we default a specific version?

The google provider is full of specific versions
https://github.com/apache/airflow/blob/a840561e1f00ce026a813039e430a085d2cfe35e/airflow/providers/google/cloud/operators/datafusion.py#L166
https://github.com/apache/airflow/blob/602abe8394fafe7de54df7e73af56de848cdf617/airflow/providers/google/marketing_platform/operators/display_video.py#L76


This means that we need to remember to bump it when the version is deprecated/old and make it as a breaking change.
Does Google offer default version that comes within its own lib?
like facebook has (for example)
https://github.com/apache/airflow/blob/602abe8394fafe7de54df7e73af56de848cdf617/airflow/providers/google/cloud/transfers/facebook_ads_to_gcs.py#L65-L66

ulsc replied:
actually, we can get rid of it altogether, because it's passed to the hook and the hook already has the version default.

eladkal replied:
i don't think removing it is the way to go.
if there are multiple versions of the API users might want to set what they wish.
My question was only about the default value of the parameter.

ulsc replied:
I understand, but anyways, if the user wants to change the api version directly from the operator, this may require a breaking change on the underlying hook. So if api_version will change, it should be changed in the hook level.

ulsc committed:
Remove `api_version` from `GoogleDriveUploadOperator`

`api_version` parameter is passed through the hook and the hook already has the api_version default. One less magic number to manage.

ulsc committed:
Fix formatting on `GoogleDriveUploadOperator` & tests

ulsc committed:
Add Google Drive Operators location to `provider.yaml`

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,115 @@
+
+    def __init__(
+        self,
+        local_paths,
```
Please add type hint

ulsc replied:
sure

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,115 @@
+        self,
+        local_paths,
+        drive_folder: str,
+        gcp_conn_id: str,
```
Can we use default conn name as we do in other Google operators?

ulsc replied:
sure

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,115 @@
+        self.delegate_to = delegate_to
+        self.impersonation_chain = impersonation_chain
+
+    def execute(self, context: Any) -> Sequence[str]:
```
```suggestion
    def execute(self, context: "Context") -> Sequence[str]:
```

You will need:
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from airflow.utils.context import Context
```

ulsc replied:
sure

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,115 @@
+                    os.remove(local_path)
+                    self.log.info(f'Deleted local file: {local_path}')
+            except FileNotFoundError:
+                self.log.error(f"File {local_path} can't be found")
```
Should this be a warning? Personally I would use error only if we would fail if the file does not exist

ulsc replied:
sure

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,115 @@
+
+        for local_path in self.local_paths:
+            self.log.info(f'Uploading file to Google Drive: {local_path}')
+            file_name = local_path.split('/')[-1]
```
I would suggest using `Path` from [pathlib](https://docs.python.org/3/library/pathlib.html) so you can do `local_path.name` instead of splitting string.

ulsc replied:
How about changing both local_paths and drive_folder to use `Path`?

ulsc committed:
Add default value to `gcp_conn_id`

ulsc committed:
Change `error` to `warning`

ulsc committed:
Comply with `pathlib.Path`

ulsc committed:
Comply tests with `pathlib.Path`

ulsc committed:
Add `TYPE_CHECKING` and conditional `Context` import

ulsc committed:
Modify documentation to comply with `pathlib.Path` arguments

ulsc committed:
Fix documentation to comply with Sphinx

turbaszek commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,118 @@
+
+    def __init__(
+        self,
+        local_paths: Sequence[Path],
```
I would also allow `Sequence[str]` to suport templating this field. I don't think jinja will resolve paths. 
```suggestion
        local_paths: Union[Sequence[Path], Sequence[str]],
```
And in constructor we need to map in case of strings

ulsc replied:
the same applies to drive_folder as well.

turbaszek reviewed (approved):
One small comment, otherwise looks good to me 🚀

potiuk committed:
Update airflow/providers/google/suite/operators/drive.py

Co-authored-by: Tomek Urbaszek <turbaszek@gmail.com>

ulsc reviewed (commented):
it would be a problem in line 104: `remote_location=str(self.drive_folder / local_path.name),` since `str` doesn't have a `name` parameter.

ulsc committed:
Add support using `Path` and `str` for jinja templating

ulsc committed:
Add types to normalized variables

ulsc committed:
Remove normalization to use `pathlib.Path` defaults to cast

ulsc committed:
Merge branch 'main' into gdrive-upload-operator

josh-fell commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,118 @@
+                self.log.warning(f"File {local_path} can't be found")
+
+        self.xcom_push(context, "remote_file_ids", remote_file_ids)
+        return remote_file_ids
```
This seems redundant. When the operator completes there will be 2 `XComs` with the same information but different keys: "return_value" and "remote_file_ids". This feels more straightforward but no strong opinion really on which one is used.
```suggestion
        return remote_file_ids
```

josh-fell commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,118 @@
+        self.delegate_to = delegate_to
+        self.impersonation_chain = impersonation_chain
+
+    def execute(self, context: "Context") -> Sequence[str]:
```
```suggestion
    def execute(self, context: "Context") -> List[str]:
```
More accurate since we know the specific sequence type being returned.

josh-fell commented on airflow/providers/google/suite/operators/drive.py:
```diff
@@ -0,0 +1,118 @@
+                    os.remove(local_path)
+                    self.log.info(f'Deleted local file: {local_path}')
+            except FileNotFoundError:
+                self.log.warning(f"File {local_path} can't be found")
```
Should this operator fail if _all_ of the files are not found? Right now the task would succeed and might give users the false sense that files were uploaded as expected. WDYT?

ulsc replied:
I think we gave the users all the necessary info with the return value. However, we may also discuss if we really want the operator to fail in any case. It may be useful if we give an error if no file_id has returned. But it may also have drawbacks, like having to include an `optional` parameter and such.

Also, in any case, checking for the availability of the local files shouldn't be in this operator's scope, and logging them is enough imo.

josh-fell replied:
I think adding a parameter related to the failure behavior here would be a good idea. If this operator succeeds but all files (or maybe even if one file?) are missing and therefore not uploaded, this is a false-positive which I think will be confusing to users. With the parameter you'd give the user an explicit option to handle this scenario. I also don't necessarily think users are always going to use the output from operators or know to look at the `XComs` pushed by the task to determine what's going on.

There is an `ignore_if_missing` parameter in some places which could be a nice analog here. WDYT?

https://github.com/apache/airflow/blob/602abe8394fafe7de54df7e73af56de848cdf617/airflow/providers/microsoft/azure/operators/wasb_delete_blob.py#L52

https://github.com/apache/airflow/blob/c1ab8e2d7b68a31408e750129592e16432474512/airflow/providers/google/cloud/operators/bigquery.py#L1784

ulsc replied:
so the next question is: when will it fail?

When there is no file_ids returned,
or
When the returned list's length is less than the local_files' length

I think, as you said it may be misleading, the latter one makes more sense. What do you think?

josh-fell replied:
I agree. If _any_ of the files are missing it makes sense to fail the task but configurable to allow files to be missing.

josh-fell commented on tests/providers/google/suite/operators/test_drive.py:
```diff
@@ -0,0 +1,59 @@
+        xcom_calls = [
+            mock.call(context, "remote_file_ids", REMOTE_FILE_IDS),
+        ]
+        mock_xcom.has_calls(xcom_calls)
```
```suggestion
        mock_xcom.assert_has_calls(xcom_calls)
```
Should use `assert_has_calls()` instead of `has_calls()` I believe based on #20453.

ulsc committed:
Update airflow/providers/google/suite/operators/drive.py

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

ulsc committed:
Update airflow/providers/google/suite/operators/drive.py

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

ulsc committed:
Update tests/providers/google/suite/operators/test_drive.py

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

ulsc committed:
Merge branch 'main' into gdrive-upload-operator

ulsc committed:
Fix imports

ulsc committed:
Merge branch 'main' into gdrive-upload-operator

ulsc committed:
Relocate the operator to transfers folder

ulsc committed:
Rename the class to `LocalFilesystemToGoogleDriveOperator`

ulsc committed:
Add documentation to `local_to_drive.rst`

ulsc committed:
Modify `provider.yaml` to comply with relocation

ulsc committed:
Relocate and rename test file

ulsc committed:
Modift tests to comply with relocation & renaming

ulsc committed:
Make title underline longer

eladkal commented on airflow/providers/google/suite/transfers/local_to_drive.py:
```diff
@@ -0,0 +1,117 @@
+                    os.remove(local_path)
+                    self.log.info(f'Deleted local file: {local_path}')
+            except FileNotFoundError:
+                self.log.warning(f"File {local_path} can't be found")
```
We should avoid f-strings in logging.
see https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

ulsc replied:
cool, wasn't aware of that.

ulsc committed:
Avoid f-strings in logging

ulsc committed:
Add `ignore_if_missing` parameter

ulsc committed:
Add link to the guide in operator description

ulsc committed:
Merge branch 'main' into gdrive-upload-operator

eladkal commented on airflow/providers/google/suite/transfers/local_to_drive.py:
```diff
@@ -0,0 +1,128 @@
+                    os.remove(local_path)
+                    self.log.info("Deleted local file: %s", local_path)
+            except FileNotFoundError:
+                self.log.warning("File %s can't be found", local_path)
```
I prefer not to be picky and catch all `OSError` but maybe that is just me :)

ulsc replied:
`FileNotFoundError` seemed like the main error cause in my case, but it'd be better to catch all `OSError`s. How about adding `OSError` after `FileNotFoundError`?

eladkal replied:
`FileNotFoundError` is subclasses of `OSError`.
So catching `OSError` is enough

eladkal replied:
https://docs.python.org/3/library/exceptions.html#os-exceptions

eladkal reviewed (approved):
LGTM
will merge when CI is green

ulsc committed:
Fix naming in tests

ulsc committed:
Add `OSError` catch & standardize logs

ulsc committed:
Merge branch 'main' into gdrive-upload-operator

ulsc committed:
Add example dags

ulsc committed:
Remove redundant import

ulsc committed:
Fix misspelling

ulsc committed:
Fix long line

ulsc committed:
Remove redundant patch from tests

eladkal commented on airflow/providers/google/suite/transfers/local_to_drive.py:
```diff
@@ -0,0 +1,130 @@
+                if self.delete:
+                    os.remove(local_path)
+                    self.log.info("Deleted local file: %s", local_path)
+            except FileNotFoundError:
+                self.log.warning("File can't be found: %s", local_path)
+            except OSError:
+                self.log.warning("An OSError occurred for file: %s", local_path)
```
merged this one too fast :) 
We should do something like:

```
        except FileNotFoundError as e:
            self.log.info('File not found: %s: %s', local_path, e)
            raise
        except OSError as e:
            self.log.info('An OSError occurred for file: %s: %s', local_path, e)
            raise
```

or

```
except OSError as e:
    raise AirflowException(
        f'Exception: There was exception related to file in path {local_path}  '
        f'full error is ({e}).'
    )
```
Would you mind open a follow up PR?

ulsc replied:
Raise would fail the dag too early, however the intention was failing after seeing the whole landscape. That's why we logged the problems and raised as a result, if user doesn't want to ignore.
